tuist/tuist
# Concurrent project generation

Part of: https://github.com/tuist/tuist/issues/1093

### Short description 📝

While profiling the workspace generator (#1093) adding concurrency to generating projects within `WorkspaceGenerator` was identified as potential improvement to gain overall performance improvements.

### Solution 📦

- Each project within a workspace is currently generated independently and non of the projects share any mutable state (the graph is static during the generation process)
- As such it is possible to run the process of generate projects within a workspace as well serialising them to disk concurrently.

### Implementation 👩‍💻👨‍💻

- [x] Add `Array+Concurrent` helpers
- [x] Add corresponding tests
- [x] Update `WorkspaceGenerator` to leverage helpers
- [x] Update `XcodeProjWriter` to leverage helpers
- [x] Update `StandardLogHandler` to ensure log lines and their terminators are written at the same time (without this change the console output had new line terminators disjoined from their corresponding lines!)
- [x] Update `TestLogHandler` to make it thread safe

### Test Plan 🛠

- Verify all automated tests pass
- Run `tuist generate` within any fixtures with multiple projects
- Verify the final console output isn't corrupted


### Profiling Results ⏱

From profiling #1093, the `WorkspaceGenerator` component could see improvements upto ~50%.


| Component                  | Old | New   | Delta     |
| --------------------- | ---- |  ----------  | --------- |
| WorkspaceGenerator  |  43.07 s  | 21.03 s     | -51.17%   |
